### ISSUE TITLE
SSL handshake fails because of too many certificate authorities

### ISSUE THREAD

**Issue by yanjunli76** [envoyproxy/envoy#14208]
*Title*: Initial ssl handshake fails because the ca cert file contains too many certificates (e.g more than 500 certificate authorities) in server side

*Description*:
During Initial ssl handshake if client authentication is enabled, envoy server will request client certificate by sending **CertificateRequest** msg\[1\]. A list of the distinguished names [X501] of certificates in ca cert file will be sent to the client. According to the rfc5246 \[1\], the maximum size of collections of distinguished names is (2^16-1) bytes. Therefore, if the number of distinguished names is more than a certain amount, the handshake will fail.

But according to the rfc5246\[1\], the handshake step by TLS/SSL server side to send the list of the distinguished names [X501] of certificates is optional. 
> If the certificate_authorities list is empty, then the client MAY send any certificate of the appropriate ClientCertificateType, unless there is some external arrangement to the contrary.

Is there a way to send an empty list in this handshake step, instead of a list of existing distinguished names of certificates?

Thanks!

\[1\]: https://tools.ietf.org/html/rfc5246#page-53

**Comment by lizan** [comment]
I don't think we have a configuration to do this. @PiotrSikora from the spec this seems valid?

A proper workaround will be reducing the number of the certificate authorities in the TLS config, instead have a common root CA signing all them as intermediate.

**Comment by PiotrSikora** [comment]
This feature request is valid per spec, although I'm not sure if all clients support it, but like @lizan said, if you're sending 500 names (those are names of the issuers, not subject names), then it seems like your signing policies and/or TLS configuration should be redesigned a bit.

**Comment by yanjunli76** [comment]
Hi @lizan @PiotrSikora, thanks for your reply.

Do you think if we can use an external Auth server to verify the client certificates, instead of verifying the client certificate by envoy during SSL handshake?

The message sent to the external Auth server can include the peer certificate (https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/attribute_context.proto#envoy-v3-api-msg-service-auth-v3-attributecontext-peer). 
I'm thinking If we only enable the one-way authentication (I assume the server will not send the list of ca authorities in the one-way authentication connection), envoy will not verify the client certificate during the TLS handshake.
Then when the envoy forwards the request to the external auth server, we can verify the client certificate against as many ca authorities as we want in the external auth server.

Do you think if this can work? Any help would be appreciated.

**Comment by xxx0624** [comment]
> Hi @lizan @PiotrSikora, thanks for your reply.
> 
> Do you think if we can use an external Auth server to verify the client certificates, instead of verifying the client certificate by envoy during SSL handshake?
> 
> The message sent to the external Auth server can include the peer certificate (https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/attribute_context.proto#envoy-v3-api-msg-service-auth-v3-attributecontext-peer).
> I'm thinking If we only enable the one-way authentication (I assume the server will not send the list of ca authorities in the one-way authentication connection), envoy will not verify the client certificate during the TLS handshake.
> Then when the envoy forwards the request to the external auth server, we can verify the client certificate against as many ca authorities as we want in the external auth server.
> 
> Do you think if this can work? Any help would be appreciated.

Hi Yanjun, did you find any way to access the client cert in the ext auth service when only enabling the one-way ssl?
Thanks!

**Comment by yanjunli76** [comment]
> > Hi @lizan @PiotrSikora, thanks for your reply.
> > Do you think if we can use an external Auth server to verify the client certificates, instead of verifying the client certificate by envoy during SSL handshake?
> > The message sent to the external Auth server can include the peer certificate (https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/attribute_context.proto#envoy-v3-api-msg-service-auth-v3-attributecontext-peer).
> > I'm thinking If we only enable the one-way authentication (I assume the server will not send the list of ca authorities in the one-way authentication connection), envoy will not verify the client certificate during the TLS handshake.
> > Then when the envoy forwards the request to the external auth server, we can verify the client certificate against as many ca authorities as we want in the external auth server.
> > Do you think if this can work? Any help would be appreciated.
> 
> Hi Yanjun, did you find any way to access the client cert in the ext auth service when only enabling the one-way ssl?
> Thanks!

Yes,  you need to set trust_chain_verification =  ACCEPT_UNTRUSTED (https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#enum-extensions-transport-sockets-tls-v3-certificatevalidationcontext-trustchainverification) and require_client_certificate = true (https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/tls.proto#envoy-v3-api-field-extensions-transport-sockets-tls-v3-downstreamtlscontext-require-client-certificate)

Once you config the ext auth service, the service should be able to get the client certificate. 

But I'm not sure how to exactly implement this. The contour has implemented this feature: https://github.com/projectcontour/contour/issues/3253 

We are using contour as our ingress controller.

**Comment by xxx0624** [comment]
> > > Hi @lizan @PiotrSikora, thanks for your reply.
> > > Do you think if we can use an external Auth server to verify the client certificates, instead of verifying the client certificate by envoy during SSL handshake?
> > > The message sent to the external Auth server can include the peer certificate (https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/auth/v3/attribute_context.proto#envoy-v3-api-msg-service-auth-v3-attributecontext-peer).
> > > I'm thinking If we only enable the one-way authentication (I assume the server will not send the list of ca authorities in the one-way authentication connection), envoy will not verify the client certificate during the TLS handshake.
> > > Then when the envoy forwards the request to the external auth server, we can verify the client certificate against as many ca authorities as we want in the external auth server.
> > > Do you think if this can work? Any help would be appreciated.
> > 
> > 
> > Hi Yanjun, did you find any way to access the client cert in the ext auth service when only enabling the one-way ssl?
> > Thanks!
> 
> Yes, you need to set trust_chain_verification = ACCEPT_UNTRUSTED (https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#enum-extensions-transport-sockets-tls-v3-certificatevalidationcontext-trustchainverification) and require_client_certificate = true (https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/tls.proto#envoy-v3-api-field-extensions-transport-sockets-tls-v3-downstreamtlscontext-require-client-certificate)
> 
> Once you config the ext auth service, the service should be able to get the client certificate.
> 
> But I'm not sure how to exactly implement this. The contour has implemented this feature: [projectcontour/contour#3253](https://github.com/projectcontour/contour/issues/3253)
> 
> We are using contour as our ingress controller.

Cool, thanks for sharing this!

**Comment by xxx0624** [comment]
Hi @yanjunli76 in my case (https://github.com/envoyproxy/envoy/issues/17348), the trusted CAs in the "Certificate Request" during the ssl handshake is empty.
So just curious about this @lizan do you have any plan to support this or is there any temporary workaround?
